### ISSUE TITLE
Feature request: conditional hails

### ISSUE THREAD

**Issue by bene-dictator** [endless-sky/endless-sky#8379]
### Problem Description

It would be useful for people who like editing and making hails (me...) to have hails that only happen when some conditions are met. Currently using events repeatedly is not feasible because it would bloat up the save file.

I see uses in:
- ships commenting on recent events (the events of the civil war, the Geminus and Martini bombings, Hai Revealed events)
- ships talking about their outfits (boasting about their weapons or outfits, telling the player to get a scram drive because they have one)
- ships changing their speech based on location (pirate ships would say that they just came from a raid on a nearby planet, not a planet halfway across human space)
- ships commenting on other hails (the bad pun hail could have other ships say stuff in response)
- ships commenting on what's happening in the system (pirate attack commenting, slow thrusters...)

While a lot of people would ignore hails, I think these would go some way into having NPC ships that appear to respond and participate in what's going on instead of telling a joke as they're bombarded by a Pirate Leviathan.

### Related Issue Links

- #6516
- #7488

### Desired Solution

For example, a hail about a merchant boasting about their new atomic engines (to go with RisingLeaf's outfit tags, #7482):
```
has "outfit tag: atomic engines" # probably not how it would look, but just thinking
  phrase "engine boast"
    word
      "I just "
      "I "
    word
      "bought "
      "purchased "
      "outfitted my ship with "
      "installed "
    word
      "these "
    word
      "snazzy "
      "lightning-quick "
      "hot "
      "amazing new "
    word
      "atomic engines"
    word
      ". Look how fast my ship goes now"
      " from the Deep"
      " from Deep Sky"
      " on Valhalla"
      ""
    word
      "!"
      "."
```
The hail would only be sent if the ship had atomic engines.

### Alternative Approaches

The problem is that what I suggest above would not fit the current syntax well and would go better with a refactor of the hail system or something like that. Another problem is that configuring hails are already very annoying because a government can only have one set of them at a time for friendly, hostile etc.

Right now it is too late for me to think of better syntax for this so I'll leave it as it is for now, just to give you all an idea of what I was thinking.

**Comment by Hurleveur** [comment]
We can already do this (sort of) using hail replacements that have conditions. That said its very finicky, so what we could do is have a `to display` optional node for phrases, simply making it so that they're not taken if the conditions are not met.

**Comment by reticent-rem** [comment]
Mentioned this in discord last week but figured I should also post here for posterity. Another use case for conditional hails/news would be to direct players to available missions in an in-universe way, so the player gets nudges towards available missions that stop appearing once those missions are done.

**Comment by marius851000** [comment]
I’m interested in trying to implementing this (this looks like a good first issue to get familiar with that engine).
I considered two different variant of structure (one using `phrase` as proposed, the other using a dedicated `hail`, like for news)

I would like to collect feedback about two way to approach this:
The first is, as proposed, to extend the `phrase` structure

```

phrase "friendly merchant hail"
	to hail
		not "event: fw armistice"
		has "chosen sides"
		# Extra condition on things like hailer or player outfit, date, etc
	word
		"Hello. I’ve heard that "
	phrase
		"band"
	word
		" took some side in the ongoing conflict"

# could add weight too.

# So...
# Downside: The data available in the condition could differ depending on the context the phrase is called. (might just ignore the condition if not an hail)
# Positive: Simpler than a dedicated "hail", and also could be usefull in other context. Condition can also be present in child (that would imply passing info needed to perform the check to the child).

# would allow to do things like

phrase "band"
	word
		"Band A"
		"Band B"
		"Band C"

phrase "band"
	to hail
		has "band z formed"
	word
		"band z"

# note that this condition might be usefull outside of just hail, but that would mean more code path to test. I’ll pass for this for now, but later might add an additional "to display" that is checked whatever the context of the phrase use is.
```

Or a second approach about a dedicated `hail` structure
(note that I had first tried to find how to make an hail that is still indexed by a text like phrases, are, but decided against as it would make it hard to add an hail from another government)

```
hail "friendly merchant hail (name could be anything)"
	to hail
		government "Merchant"
		enabled # or not disabled?
		friendly
		not "event: fw armistice"
		has "chosen sides"
	weight
		4
	message
		word
			"Hello. I’ve heard that "
		phrase
			"band"
		word
			" took some side in the ongoing conflict"
```
(this one is a bit verbose IMO. And harder to automatically check that everyone has at least an hail available no matter what)

My first iteration, in case it is usefull:
```
hail "friendly merchant hail"
	to show
		not "event: fw armistice"
		has "chosen sides"
		# This might make it harder to implement, but condition on current location, hailer government, player government, location government, player outfit, hailer outfit, hailer ship, etc might be possible
	weight # for good measure and more control over message frequency. To showcase the benefit of using a dedicated hail structure
		4
	message
		word
			"Hello. I’ve heard that "
		phrase
			"band"
		word
			" took some side in the ongoing conflict"

# Note that it might be possible to use governement filter so to not have to declare a phrase/hail to use in the government file (like for news). Low priority for now.
# For backward compatibility, wrap phrases has a matching ID into an hail structure (at the game-engine level), with a structure similar to that:
hail "friendly merchant hail"
	# no "to show": always show
	weight
		# number of phrases in here × base weight (100?)
	message
		phrase
			"friendly merchant hail"

# Problem: how to display other hail? (with phrase, it’s easy, you can just reference another phrase (you’ll need to still pass the condition. Or not?)) Potential idea:

hail "friendly pirate merchant hail"
	weight
		100
	hail
		"friendly merchant hail" # that could recursion that would need to be checked for


# Hmm... All of that is complex. How about the idea of just adding condition to phrases...

# second proposition
phrase "friendly merchant hail"
	to hail
		not "event: fw armistice"
		has "chosen sides"
		# Same things about more condition
	word
		"Hello. I’ve heard that "
	phrase
		"band"
	word
		" took some side in the ongoing conflict"
```

**Comment by marius851000** [comment]
Well, in the end, I think the phrase approach (as proposed) is the best. I’m still unsure if it’s better to keep the "to hail" or to rename it to "to display" and add a "hailing" check-able condition.

I think it is. (not that I would implement check in other context. For now)

**Comment by marius851000** [comment]
(It appears this would indeed be quite simple to implement, and that adding weight to phrases would also be easy, thought I won’t do that here)

**Comment by marius851000** [comment]
I can now do things like
```
phrase "debug hail"
	to use
		filter hailingShip
			government "Hai"
	word
		"debug hail: @@@@@@@"
```

I wonder if I should make a dedicated `hail` type like there is a dedicated `news` type

**Comment by petervdmeer** [comment]
> I wonder if I should make a dedicated `hail` type like there is a dedicated `news` type

Yes, I think that is needed; phrases are currently just blocks of text, and they appear to be used in [standard dialogs](https://github.com/endless-sky/endless-sky/blob/74218387d1546c84ef4dd4a5abd0c0f5b90d3147/data/dialog%20phrases.txt), [news](https://github.com/endless-sky/endless-sky/wiki/CreatingNews) and in **hails**. I feel that it is very appropriate to keep `phrases` as blocks of text, without hail-specific functionality, and without things like conditions to filter them.

Hails themselves are currently phrases, with a friendly and hostile variant per government. This does not support things like ship-specific hails and location specific hails, and I don't see an easy way to add such functionality to the government section.

So if we want to support conditional hails, then it makes sense to define a new `"hails"` root-node, similar to the `"news"` root-node. Something like:

```
hails <name>
	location
		<location filter that determines where (in which systems) those hails are used>
	to show
		<conditionset that determines if the hail are to be used / shown at all>
	ships
		<ships filter that determines which ships should use those hails>
	message
		<message specification composed of phrases and/or words, similar to news)
```

[Ship filters](https://github.com/endless-sky/endless-sky/issues/10041) are currently not existing ingame, and I strongly suggest to use separate PRs for the new `hails` keyword and for ship filters (so any PR that adds the `hails` keyword should only add `to show` and `message`, keep the `ships` keyword and `location` keyword for separate PRs.) Trying to do everything in a single PR will likely result in a very large PR that would not be mergeable for many years (if mergeable at all)
